### PR TITLE
Gracefully terminate daemon goroutines

### PR DESCRIPTION
**Pull request by adrianchiris**
- Extend watcher Run() to return a Stop function to terminate the
  routines that listen to k8s objects
- Modify daemon periodic task execution to use wait.Until()
- Add logic to terminate tasks on os signals SIGINT, SIGTERM in daemon RUN()

### CODE REVIEW

**Comment by adrianchiris** [comment]
Still need to test on a real setup

**Comment by moshe010** [comment]
you have lint failures
pkg/daemon/daemon.go:31:62: `recieved` is a misspelling of `received` (misspell)
	// Execute Daemon loop, returns when os.Interrupt signal is recieved
	                                                            ^
pkg/daemon/daemon.go:122:14: `Recieved` is a misspelling of `Received` (misspell)
		glog.Info("Recieved signal %s. Terminating...", sig)
		           ^
pkg/daemon/daemon.go:120:2: S1000: should use a simple channel send/receive instead of `select` with a single case (gosimple)
	select {
	^
pkg/daemon/daemon.go:118:15: SA1017: the channel used with signal.Notify should be buffered (staticcheck)
	signal.Notify(sigChan, os.Interrupt)
	             ^
Makefile:79: recipe for target 'lint' failed

**Comment by adrianchiris** [comment]
handling SIGTERM signal was added as well as k8s sends that to pods on pod termination to allow them to gracefully shut down.

https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -101,10 +104,24 @@ func NewDaemon() (Daemon, error) {
...
+	// closing the channel will stop the goroutines executed in the wait.Until() calls below
+	stopPeriodicsChan := make(chan struct{})
+	// Run periodic tasks
+	go wait.Until(d.AddPeriodicUpdate, time.Duration(d.config.PeriodicUpdate)*time.Second, stopPeriodicsChan)
+	go wait.Until(d.DeletePeriodicUpdate, time.Duration(d.config.PeriodicUpdate)*time.Second, stopPeriodicsChan)
 	glog.Info("watcher Run():")
-	d.watcher.Run()
+	// Running Watcher
+	// calling watcherStopFunc() will stop the watcher which listens to events on a separate goroutine
+	watcherStopFunc := d.watcher.Run()
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: you are hiding the go routines in the watcher.run which is confusing even with the comment.  Why not go d.watcher.Run()?
also can we use one channel to close all go routines by passing it in the d.watcher.Run?  (and remove the watcherStopFunc)

File: pkg/watcher/watcher.go [hunk 2]
```diff
@@ -12,29 +12,34 @@ import (
...
 	watchList := cache.NewListWatchFromClient(client.GetRestClient(), resource, kapi.NamespaceAll, fields.Everything())
-	return &watcher{eventHandler: eventHandler, watchList: watchList, stopChan: make(chan struct{})}
+	return &watcher{eventHandler: eventHandler, watchList: watchList}
 }
 
-// Run for for infinite listening for k8s resource events
-func (w *watcher) Run() {
+// Run until StopFunc is called, listening for k8s resource events
+func (w *watcher) Run() StopFunc {
+	stopChan := make(chan struct{})
```
Comment (moshe010) [inline pkg/watcher/watcher.go]: why get the channel from the caller and let him close it?

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -101,10 +104,24 @@ func NewDaemon() (Daemon, error) {
 
 func (d *daemon) Run() {
 	glog.Info("daemon Run():")
-	go wait.Forever(d.AddPeriodicUpdate, time.Duration(d.config.PeriodicUpdate)*time.Second)
-	go wait.Forever(d.DeletePeriodicUpdate, time.Duration(d.config.PeriodicUpdate)*time.Second)
+	// closing the channel will stop the goroutines executed in the wait.Until() calls below
+	stopPeriodicsChan := make(chan struct{})
+	// Run periodic tasks
+	go wait.Until(d.AddPeriodicUpdate, time.Duration(d.config.PeriodicUpdate)*time.Second, stopPeriodicsChan)
+	go wait.Until(d.DeletePeriodicUpdate, time.Duration(d.config.PeriodicUpdate)*time.Second, stopPeriodicsChan)
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: defer close(stopPeriodicsChan)

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (moshe010) [reply pkg/daemon/daemon.go]: ok let do this, defer watcherStopFunc()

Comment (moshe010) [reply pkg/daemon/daemon.go]: so let change Run to RunInBackground

**Pull request merged**